### ISSUE TITLE
Calculating gradients only for the selected parameters

### ISSUE THREAD

**Issue by cagrikymk** [google/jax#4752]
I have a unique model that consists of arrays with different shapes. I only want to optimize a subset of this model. Instead of doing masked updates, I decided to use some kind of parameter selection procedure which forces jax to calculate the gradients only for the selected parameters and indices. By doing so, there is no need to do any post-processing for grads or hessian matrix. 

'use_selected_parameter` function is breaking the pure function rule of jax (I think) but it nicely does what I am trying to achieve. 
However, sometimes I get an error related to tracers.

I can carry this selection logic outside of the "jax.grad" region but I wonder if there is a better way of doing this. 

My approach is similar to this minimal example I created:

```
import jax
import jax.numpy as jnp
import numpy as onp

def create_model(sizes):
	model = []
	for shape in shapes:
		model.append(jnp.array(onp.random.normal(size=shape)))
	return model

def use_selected_params(selected_params,selected_indices,model):
	for i, ind in enumerate(selected_indices):
		model[ind[0]] = jax.ops.index_update(model[ind[0]], ind[1], selected_params[i])
	return model

def calculate_loss(selected_params,selected_indices,model):
	model = use_selected_params(selected_params,selected_indices,model)

	summ = jnp.sum([jnp.sum(p) for p in model])

	return summ


shapes = [(20,),(20,20), (20,20,20)]
my_model = create_model(shapes)

selected_params = jnp.array(onp.random.normal(size=5))
selected_indices = [(0,1), (0,2), (1, (3,4)), (2, (3,4,5))]

loss = calculate_loss(selected_params,selected_indices,my_model)

# jit works
val_and_grad_func = jax.jit(jax.value_and_grad(calculate_loss),static_argnums=(1,))
grads = val_and_grad_func(selected_params,selected_indices,my_model)

LR = 0.5
loss_vals = []

# optimizer
for i in range(10):
	loss,grads = val_and_grad_func(selected_params,selected_indices,my_model)
	selected_params = selected_params - LR * grads
	loss_vals.append(loss)



# without jit
val_and_grad_func = jax.value_and_grad(calculate_loss)
grads = val_and_grad_func(selected_params,selected_indices,my_model)

LR = 0.5
loss_vals2 = []

# optimizer
for i in range(10):
	loss,grads = val_and_grad_func(selected_params,selected_indices,my_model)
	selected_params = selected_params - LR * grads
	loss_vals2.append(loss)

#Error: Encountered an unexpected tracer.
#	Perhaps this tracer escaped through global state from a previously traced function
```
